design/k8s_secrets_labels_design.md

```diff
@@ -0,0 +1,218 @@
+
+## Background
+
+As part of Phase 2 for Secrets Provider for K8s effort, we want to improve the customer experience by offering the ability to use labels for easy identification of K8s secrets whose secret values should come from Conjur. This will allow customers to list in terms of groupings instead of having to list the K8s secrets individually.
+
+## Issue description
+
+At current, the Secrets Provider code looks at the `K8S_SECRETS` environment variable in each of the pod manifests to know which K8s Secret to update with Conjur values. 
+
+Listing each K8s Secret in this way is not a scalable solution, especially for customers with hundreds of K8s Secrets. 
+
+## Solution
+
+We will offer the customer an *additional* option to define which K8s Secrets values are hosted in Conjur by defining them in terms of a label instead of individually. They will define the label that is attached to the K8s Secret and add it under `K8S_SECRETS_LABEL` environment variable in the Secrets Provider manifest. 
+
+The Secrets Provider will then make a separate API call to [K8s](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#secret-v1-core)/[OC](https://docs.openshift.com/container-platform/4.4/rest_api/index.html#list-14), fetching all K8s Secrets with that label. We will use that information to fetch their values in Conjur and update the K8s Secrets.
+
+This enhancement can be used in the current *init solution* and will help build the foundation for later Milestones.
+
+### Design
+
+|      | Fetch/Update K8s Secret                                      | Pros                                                         | Cons                                                         | UX                                                           | Effort estimation |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ----------------- |
+| 1    | Get K8s secrets with the names specified in `K8S_SECRETS`    | - No code changes  - Backwards compatible                    | - Repetitive work <br />- Not scalable                       | List K8S Secrets one-by-one                                  | Free              |
+| 2    | Read all K8s Secrets that are marked with labels specified in `K8S_SECRETS_LABEL` env var | - Scalable grouping  - Less intervention on our part as a customer can define which types of secrets we should parse  - K8s native, as-is solution | - Not general solution so we are still K8s bound - Demands SA has "list" privileges on secrets | Add "list" to SA `verbs: ["get", "update", "list"]` <br />- List label grouping in Secrets Provider manifest<br />- Label key `conjur-secrets-label` will need to be added to each K8s Secret | 5 days            |
```

Comment:
> Label key `conjur-secrets-label` will need to be added to each K8s Secret

only if they don't already have a label they want to use

Reply:
So we are evaluating a few different directions
1. Supply a key/value pair that the customer cannot configure. Ex: `conjur-secrets-label: true`
2. Supply a `conjur-secrets-label` key that the customer cannot configure but a configurable value. For example: `conjur-secrets-label: prod`
3. Allow the customer to configure both multiple key/value pairs. This is will be problematic for upcoming milestones when we will have to monitor changes in the K8s environment by performing `watch` on all k8s secrets resources. When querying the K8s Watch endpoint, the customer would supply a [comma-separated list which acts as &&](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors). So the following:
```
key: K8S_SECRETS_LABELS
value: config-label1:conjur-key1, config-label2:conjur-key2`
```
would mean that the K8S Secret that would be returned would need to have `config-label1:conjur-key1` **&&** `config-label2:conjur-key2`

_A way to curb this_ would be limit the user to a single key and multiple values. For example:  
 `my-config-conjur-label in (prod, dev)`. K8s gives us the option to use [set-based requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement) which gives us OR relationships on the values

```
key: K8S_SECRETS_LABELS
value: my-config-conjur-label in (production, dev)'
```

Reply:
In case you go with option 2 then i would change the value of `K8S_SECRETS_LABELS` to have only the values.
if `conjur-secrets-label` is hard-coded then they don't need to add it all the time.

Instead of 
```
key: K8S_SECRETS_LABELS
value: conjur-secrets-label:prod,conjur-secrets-label:staging
```

do
```
key: K8S_SECRETS_LABELS
value: prod,staging
```

Reply:
but i really think that this is a complicated solution. I would go with option 1 and let the customer decide if they want segregation of duty between secrets-providers (and then define the secrets by name) or have all the secrets-providers get all the secrets (and then get them by the static label `conjur-secrets-label:true` without the need to define the env var `K8S_SECRETS_LABELS`).